concordium-consensus/src/Concordium/KonsensusV1/Consensus/Quorum.hs

```diff
@@ -0,0 +1,283 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- |This module contains the logic for creating 'QuorumCertificate's, receiving and
+-- verifying quorum signatures that is used for the consensus v1 protocol.
+module Concordium.KonsensusV1.Consensus.Quorum where
+
+import Control.Monad.Catch
+import Control.Monad.State
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import Lens.Micro.Platform
+
+import Concordium.Genesis.Data.BaseV1
+import Concordium.Logger
+import Concordium.TimeMonad
+import Concordium.Types
+import Concordium.Types.Parameters
+import Concordium.Utils
+
+import Concordium.GlobalState.BlockState
+import qualified Concordium.GlobalState.Persistent.BlockState as PBS
+import qualified Concordium.GlobalState.Types as GSTypes
+import Concordium.KonsensusV1.Consensus
+import Concordium.KonsensusV1.Consensus.Finality
+import Concordium.KonsensusV1.Flag
+import Concordium.KonsensusV1.TreeState.Implementation
+import qualified Concordium.KonsensusV1.TreeState.LowLevel as LowLevel
+import Concordium.KonsensusV1.TreeState.Types
+import Concordium.KonsensusV1.Types
+
+-- |Reasons that a 'QuorumMessage' can be rejected.
+data ReceiveQuorumMessageRejectReason
+    = -- |The 'Round' presented in the 'QuorumMessage' is obsolete.
+      ObsoleteRound
+    | -- |The finalizer for the 'QuorumMessage' is not present
+      -- in the finalization committee.
+      NotAFinalizer
+    | -- |The signature on the 'QuorumMessage' is invalid.
+      InvalidSignature
+    | -- |The finalizer has already signed another 'QuorumMessage' for the
+      -- 'Round'.
+      AlreadySigned
+    | -- |The 'QuorumMessage' points to an invalid block.
+      InvalidBlock
+    | -- |The 'Round' of the 'QuorumMessage' and the 'Round' of the
+      -- block that it points to are not consistent.
+      InconsistentRounds
+    | -- |The 'Epoch' of the 'QuorumMessage' and the 'Epoch' of the
+      -- block that it points to are not consistent.
+      InconsistentEpochs
+    deriving (Eq, Show)
+
+-- |Result codes for receiving a 'QuorumMessage'.
+data ReceiveQuorumMessageResult (pv :: ProtocolVersion)
+    = -- |The 'QuorumMessage' was received i.e. it passed verification.
+      Received !(VerifiedQuorumMessage pv)
+    | -- |The 'QuorumMessage' was rejected.
+      Rejected !ReceiveQuorumMessageRejectReason
+    | -- |The 'QuorumMessage' points to a round which indicates a catch up is required.
+      CatchupRequired
+    | -- |The 'QuorumMessage' is a duplicate.
+      Duplicate
```

Comment:
I don't recall anymore - I've put the duplicate into the `ReceiveQuorumMessageRejectReason` type now.

Reply:
The same thing was done for `ReceiveTimeoutMessageResult`, the `Duplicate` case is separate from the other `Rejected` types. I didn't comment it there, because I didn't expect it to be changed. I thought I would get an explanation here, and there some would apply to the timeouts.

Reply:
Ah thanks; an oversight from my side. 

I have done the corresponding change for timeout handling as well.